### DISCUSSION TITLE
syntax highlighting support for languages with pattern matching

### DISCUSSION THREAD

**Discussion by ananthakumaran** [tree-sitter/tree-sitter#1096] [Q&A]
I am working on a [grammar](https://github.com/ananthakumaran/tree-sitter-elixir) for [Elixir](https://hexdocs.pm/elixir/syntax-reference.html) programming language and seems to have hit couple of road blocks

1.  Elixir has very good support for pattern matching. Pattern matching can be performed on multiple data types like list, tuple, map, string etc. This can be nested to any level.

```elixir
{[%{:key => "Mr. " <> prefix}]} = {[%{:key => "Mr. x"}]}
```

```text
program [0, 0] - [1, 0])
  binary_op [0, 0] - [0, 56])
    left: tuple [0, 0] - [0, 31])
      list [0, 1] - [0, 30])
        map [0, 2] - [0, 29])
          binary_op [0, 4] - [0, 28])
            left: atom [0, 4] - [0, 8])
              atom_literal [0, 4] - [0, 8])
            right: binary_op [0, 12] - [0, 28])
              left: string [0, 12] - [0, 18])
                string_start [0, 12] - [0, 13])
                string_content [0, 13] - [0, 17])
                string_end [0, 17] - [0, 18])
              right: identifier [0, 22] - [0, 28])
    right: tuple [0, 34] - [0, 56])
      list [0, 35] - [0, 55])
        map [0, 36] - [0, 54])
          binary_op [0, 38] - [0, 53])
            left: atom [0, 38] - [0, 42])
              atom_literal [0, 38] - [0, 42])
            right: string [0, 46] - [0, 53])
              string_start [0, 46] - [0, 47])
              string_content [0, 47] - [0, 52])
              string_end [0, 52] - [0, 53])
```

If I want to highlight identifiers on the left of an assignment, it's not easy to do as there is no easy way to say something like

```lisp
(binary_op
 left: (* (identifier) @variable)
 operator: "=")
```

This is already raised in other [threads](https://github.com/tree-sitter/tree-sitter/discussions/981)

2.  Second issue is kind of similar to first. In Elixir many of the features like module, function definition etc are built on top of macro. At AST level everything is represented as a function call. To highlight function parameter we use something like below<sup><a id="fnr.1" class="footref" href="#fn.1">1</a></sup>.

```lisp
(call function: (function_identifier) @keyword
    [(call
      function: (function_identifier) @function
      (arguments
       [(identifier) @variable.parameter
        (_ (identifier) @variable.parameter)
        (_ (_ (identifier) @variable.parameter))
        (_ (_ (_ (identifier) @variable.parameter)))
        (_ (_ (_ (_ (identifier) @variable.parameter))))
        (_ (_ (_ (_ (_ (identifier) @variable.parameter)))))]))
     (binary_op
      left:
      (call
       function: (function_identifier) @function
       (arguments
        [(identifier) @variable.parameter
         (_ (identifier) @variable.parameter)
         (_ (_ (identifier) @variable.parameter))
         (_ (_ (_ (identifier) @variable.parameter)))
         (_ (_ (_ (_ (identifier) @variable.parameter))))
         (_ (_ (_ (_ (_ (identifier) @variable.parameter)))))]))
      operator: "when")
     (binary_op
      left: (identifier) @variable.parameter
      operator: _ @function
      right: (identifier) @variable.parameter)]
    (#match? @keyword "^(defp|def|defmacrop|defmacro|defguardp|defguard|defdelegate)$")
    (#match? @variable.parameter "^[^_]"))
```

The issue here is, given a function like the one below, it stops highlighting after 5<sup>th</sup> argument.

```elixir
def func3(arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9)  
```
```
program [0, 0] - [1, 0])
  call [0, 0] - [0, 63])
    function: function_identifier [0, 0] - [0, 3])
    call [0, 4] - [0, 63])
      function: function_identifier [0, 4] - [0, 9])
      arguments [0, 9] - [0, 63])
        identifier [0, 10] - [0, 14])
        identifier [0, 16] - [0, 20])
        identifier [0, 22] - [0, 26])
        identifier [0, 28] - [0, 32])
        identifier [0, 34] - [0, 38])
        identifier [0, 40] - [0, 44])
        identifier [0, 46] - [0, 50])
        identifier [0, 52] - [0, 56])
        identifier [0, 58] - [0, 62])
```

## Footnotes

<sup><a id="fn.1" class="footnum" href="#fnr.1">1</a></sup> : <https://github.com/ananthakumaran/tree-sitter-elixir/blob/master/queries/highlights.scm>

**Comment by alemuller** [comment]
Maybe tree-sitter could provide a `#has-parents?` and `#has-ancestors?` predicated like nvim-tree-sitter? 

> The issue here is, given a function like the one below, it stops highlighting after 5th argument.

I've stumble upon a similar problem in the past. But instead of nested parameters, it was a repetition at same hierarchy.

**Reply by ananthakumaran** [reply]
> But instead of nested parameters, it was a repetition at same hierarchy.

the second issue is about repetition at the same level, not nesting, I added the AST to make it clear. The query actually tries to handle nesting up to 5 levels, but syntax highlighting fails even without nesting
